packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,267 @@ import { Far } from '@endo/marshal';
+        { durable: true },
+      ),
+      nextAddressNumber: 1n,
+      // This assumes one connection per host.
+      // If that is not the case, receiver should instead be a
+      // durable makeScalarBigWeakMapStore keyed by connection.
+      receiver: null,
+      remoteLocatorUnum: null,
+    };
+  };
+  const makeNetworkHost = defineDurableKindMulti(
+    networkHostHandle,
+    initNetworkHostState,
+    {
+      host: {
+        publish: ({ state }, obj) => {
+          const address = `/alleged-name/${state.allegedName}/egress/${state.nextAddressNumber}`;
+          state.nextAddressNumber += 1n;
+          state.exportedObjects.init(address, obj);
+          return address;
+        },
+        lookup: ({ state }, address) => {
+          return E(state.remoteLocatorUnum).lookup(address);
+        },
+      },
+      commsSetReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.receiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.receiver = receiver;
+        },
+      },
+      commsTransmitter: {
+        transmit: ({ state }, msg) => {
+          // 'msg' will be a string (vats/comms/outbound.js deliverToRemote)
+          E(state.connection).send(msg);
+        },
+      },
+      localLocatorUnum: {
+        lookup: ({ state }, address) => {
+          return state.exportedObjects.get(address);
+        },
+      },
+      handler: {
+        onOpen: async ({ state, facets }, connection, ..._args) => {
+          const name = state.connectionName;
+          assert(!state.connection, X`host ${name} already opened`);
+          state.connection = connection;
+
+          const comms = state.comms;
+          // TODO: Can these calls be made in parallel?
+          await E(comms).addRemote(
+            name,
+            facets.commsTransmitter,
+            facets.commsSetReceiver,
+          );
+          await E(comms).addEgress(name, 0, facets.localLocatorUnum);
+          state.remoteLocatorUnum = await E(comms).addIngress(name, 0);
+        },
+        onReceive: ({ state }, _connection, msg) => {
+          // setReceiver ought to be called before there's any chance of
+          // onReceive being called
+          E(state.receiver).receive(msg);
+        },
+        onClose: ({ state }, _connection, ..._args) => {
+          state.receiver = null;
+          // TODO: What would it take to fully support this?
+          console.warn(`deleting connection is not fully supported in comms`);
+        },
+        infoMessage: ({ state }, ...args) => {
+          void E(state.console).log('VatTP connection info:', ...args);
+        },
+      },
+    },
+  );
+
+  // Expose a durable service singleton.
+  return vivifySingleton(baggage, serviceSingletonBaggageKey, {
+    ...serviceMailboxFunctions,
+    makeNetworkHost: (allegedName, comms, cons = console) => {
```

Comment:
tiny nit, this combination of a `...` splat and a single name initially made me think that `makeNetworkHost` was somehow overriding one portion of `serviceMailboxFunctions`, and it took me a minute to realize this was composing two separate bits of functionality into the return value

what do you think about defining `networkFunctions = { makeNetworkHost }` and then having this be `{ ...serviceMailboxFunctions, ...networkFunctions }` ?

but it's fine either way

Reply:
(retroactive review note for future historians: it turns out that returning a virtual/durable object from `buildRootObject` causes problems, and I missed this during review. #7240 has details, and our PR to prohibit vats from doing this also changes this line to use `Far` instead of `vivifySingleton`)